SOSS/dms/soss_centroids.py

```diff
@@ -0,0 +1,377 @@
+        _plot_centroid(image_masked, xtrace, ytrace)
+
+    return xtrace, ytrace, param
+
+
+def test_uncontam_centroids():
+    """"""
+
+    filename = os.path.join(PATH, 'SOSS_ref_2D_profile_SUBSTRIP256.fits.gz')
+
+    image, header = fits.getdata(filename, ext=2, header=True)
+    xtrace, ytrace, param = get_uncontam_centroids(image, header=header, verbose=True)
+
+    return
+
+
+def edge_trigger(image, halfwidth=5, yos=1, verbose=False):
+    """Detect the edges and center of the trace based on the minima and maxima of the derivate
+     of the columns, which is computed in a running window along the columns of the detector image
+
+     :param image: A 2D image of the detector.
+     :param halfwidth: the size of the window used when computing the derivatives.
+     :param yos: the oversampling factor of the image array along the y-direction.
+     :param verbose: If set True some diagnostic plots will be made.
+
+     :type image: array[float]
+     :type halfwidth: int
+     :type yos: int
+     :type verbose: bool
+
+     :returns: ytrace_max, ytrace_min, ytrace_comb - The upper edge, lower edge and center of the trace.
+     :rtype: Tuple(array[float], array[float], array[float])
+     """
+
+    dimy, dimx = image.shape
+    halfwidth = halfwidth * yos
+
+    # Create coordinate arrays.
+    xpix = np.arange(dimx)
+    ypix = np.arange(dimy)
+    _, ygrid = np.meshgrid(xpix, ypix)
+
+    # Compute windowed slopes over the columns.
+    slopevals = np.zeros_like(image)
+    for irow in range(halfwidth, dimy-halfwidth):
+
+        # Compute the window indices.
+        ymin = irow - halfwidth
+        ymax = irow + halfwidth + 1
+
+        # Get the x and y data to find the slope to.
+        datay = image[ymin:ymax, :]
+        mask = np.isfinite(datay)
+        datax = np.where(mask, ygrid[ymin:ymax, :], np.nan)  # Need to set values NaN in y to NaN in x.
+
+        # Compute the slope.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            xmean = np.nanmean(datax, axis=0, keepdims=True)
+            ymean = np.nanmean(datay, axis=0, keepdims=True)
+
+        with np.errstate(invalid='ignore'):
+            slope = np.nansum((datax - xmean) * (datay - ymean), axis=0) / np.nansum((datax - xmean) ** 2, axis=0)
+
+        # Set slopes computed from < 3 datapoints to NaN.
+        slopevals[irow, :] = np.where(np.sum(mask, axis=0) >= 3, slope, 0.)
+
+    # Find the upper and lower bounds on the trace.
+    args = np.nanargmax(slopevals, axis=0)
+    vals = np.nanmax(slopevals, axis=0)
+    ytrace_max = np.where(vals != 0, ypix[args], np.nan)
+
+    args = np.nanargmin(slopevals, axis=0)
+    vals = np.nanmin(slopevals, axis=0)
+    ytrace_min = np.where(vals != 0, ypix[args], np.nan)
+
+    # Scan through a range of trace widths.
+    slopes_best = np.zeros_like(xpix)
+    ytrace_best = np.zeros_like(xpix)
+    widths_best = np.zeros_like(xpix)
+    for width in range(18*yos, 27*yos):
```

Comment:
spaces between operators: `range(18 * yos, 27 * yois)`

Reply:
I've always preferred spacing around addition and subtraction but not multiplication it looks more natural to me. Like how it'd look in say Latex.